reports-backend/src/main/java/org/jboss/da/reports/impl/ReportsGeneratorImpl.java

```diff
@@ -369,4 +549,66 @@ private ProductArtifact toProductArtifact(ProductVersionArtifactRelationship pa)
         return ret;
     }
 
+    @Override
+    public List<LookupReport> getLookupReportsForGavs(LookupGAVsRequest request) {
+        
+        // Use a concurrent list since we're using parallel stream to add stuff to the list
+        CopyOnWriteArrayList<LookupReport> reportsList = new CopyOnWriteArrayList<>();
+
+        boolean communicationSucceeded = request.getGavs().parallelStream().distinct().map((gav) -> {
+            try {
+                VersionLookupResult lookupResult = versionFinder.lookupBuiltVersions(gav);
+                LookupReport lookupReport = toLookupReport(gav, lookupResult);
+                reportsList.add(lookupReport);
+                return true;
+            } catch (CommunicationException ex) {
+                log.error("Communication with remote repository failed", ex);
+                return false;
+            }
+        }).allMatch(x -> {
+            return x;
+        });
+
+        if(communicationSucceeded){
+            Set<ProductVersion> relevantProductVersions = getProductVersions(request.getProductNames(), request.getProductVersionIds());
+            Map<WhiteArtifact, Set<ProductVersion>> whiteArtifactsAndOrigin = getWhiteArtifacts(relevantProductVersions);
+            
+            reportsList.stream().forEach(x -> 
+                {
+                    GAV currentGav = x.getGav();
+                    List<String> availableVersionsFromWhitelist = getAvailableWhitelistVersions(whiteArtifactsAndOrigin, currentGav);
+                    String bestVerStr = x.getBestMatchVersion();
+                    Optional<String> bestVersion = getBestMatchVersionFromWhitelist(
+                            bestVerStr == null ? Optional.empty() : Optional.of(bestVerStr), 
+                            currentGav, availableVersionsFromWhitelist
+                    );
+                    
+                    // Combine existing with whitelist versions
+                    if(x.getAvailableVersions() != null){
+                        availableVersionsFromWhitelist.addAll(x.getAvailableVersions());
+                    }
+                    x.setAvailableVersions(availableVersionsFromWhitelist);
+                    x.setBestMatchVersion(bestVersion.orElse(null));
+                }
+             );
+            
+            return reportsList;
+        }
+
+        return null;
```

Comment:
What about throwing CommunicationException or returning Optional from this method, instead of returning null?